File: 7-Grafos/1082-Componentes-Conexos/1082.cpp
#include <iostream>
#include <vector>
#include <algorithm>

int dfs (const std::vector<std::vector<int>>& graph, std::vector<std::vector<int>>& connected, int V) {
    std::vector<bool> visited(V, false);
    std::vector<int> stack;
    int node, index = -1;
    for (int j = 0; j < V; ++j) {
        if (!visited[j]) {
            ++index;
            stack.push_back(j);
            while (stack.size() > 0) {
                node = stack[stack.size()-1];
                stack.pop_back();
                if (!visited[node]) {
                    connected[index].push_back(node);
                    visited[node] = true;
                    for (int neighbor : graph[node]) {
                        stack.push_back(neighbor);
                    }
                }
            }
        }
    }
    return index;
}

int main() {
    int V, E, cases, index;
    char u[2], v[2];
    std::cin>>cases;
    for (int i = 0; i < cases; ++i) {
        std::cin>>V>>E;
        std::vector<std::vector<int>> graph(V);
        std::vector<std::vector<int>> connected(V);
        for (int j = 0; j < E; ++j) {
            std::cin>>u>>v;
            u[0] -= 97; v[0] -= 97;
            graph[u[0]].push_back(v[0]);
            graph[v[0]].push_back(u[0]);   
        }
        index = dfs(graph, connected, V);
        std::cout<<"Case #"<<i+1<<":\n";
        for (int j = 0; j <= index; ++j) {
            std::sort(connected[j].begin(), connected[j].end());
            for (int el : connected[j]) {
                std::cout<<(char)(el+97)<<",";
            }
            std::cout<<"\n";
        }
        std::cout<<index+1<<" connected components\n\n";
    }
    return 0;
}

File: 7-Grafos/1082-Componentes-Conexos/1082.py
def dfs(graph):
    visited, connected, stack = [], [], []
    index = -1
    for node in graph:
        if node not in visited:
            connected.append([])
            index += 1
            stack.append(node)
            while stack:
                v = stack.pop()
                if v not in visited:
                    connected[index].append(v)
                    visited.append(v)
                    stack.extend(reversed(graph[v]))
    return connected

alphabet = ['a', 'b', 'c', 'd', 'e', 'f', 'g', 'h', 'i', 'j', 'k', 'l', 'm', 'n', 'o', 'p', 'q', 'r', 's', 't', 'u', 'v', 'w', 'x', 'y', 'z']
for i in range(int(input())):
    graph = {}
    nvertices, nedges = input().split()
    nvertices, nedges = int(nvertices), int(nedges)
    graph.update({alphabet[i] : [] for i in range(nvertices)})
    for q in range(nedges):
        edge, neighbor = input().split()
        graph[edge].append(neighbor)
        graph[neighbor].append(edge)
    print (f"Case #{i+1}:")
    counter = 0
    for connected in dfs(graph):
        connected.sort()
        print (',' .join(connected), end='')
        print (',')
        counter += 1
    print (counter, "connected components\n")